main.cpp

```diff
@@ -270,101 +219,165 @@ struct HeapA
+        value.reset(nullptr);
+    }
 
-struct DoubleType;
-struct IntType;
+    Numeric& operator+=( Type rhs )
+    {
+        *value += rhs; 
+        return *this;
+    }
+    Numeric& operator-=( Type rhs )
+    {
+        *value -= rhs; 
+        return *this;
+    }
+    Numeric& operator*=( Type rhs )
+    {
+        *value *= rhs; 
+        return *this;
+    }
 
-struct FloatType
-{
-    explicit FloatType(float val) : value(new float(val)) {}
-    ~FloatType()
+    template<typename ParamType>
+    Numeric& operator/=( ParamType rhs )
     {
-        delete value;
+        if constexpr (std::is_same<Type, int>::value)
+        {
+            if constexpr (std::is_same<ParamType, int>::value)
+            {
+                if(rhs == 0)
+                {
+                    std::cout << "error: integer division by zero is an error and will crash the program!\n";
+                    return *this;
+                }
+            }
+            else if (std::abs(rhs) <= std::numeric_limits<T>::epsilon())
+            {
+                std::cout << "can't divide integers by zero!\n";
+                return *this;
+            }
+        } 
+        else if (std::abs(rhs) <= std::numeric_limits<T>::epsilon())
+        {
+            std::cout << "warning: floating point division by zero!" << std::endl;  
+        }
+       
+        *value /= rhs;
+        return *this;
     }
 
-    FloatType& operator+=( float rhs );
-    FloatType& operator-=( float rhs );
-    FloatType& operator*=( float rhs );
-    FloatType& operator/=( float rhs );
+    operator Type() const { return *value; }
 
-    operator float() const { return *value; }
+    //Numeric& pow(Type t); 
+    Numeric& pow(const Type& t)
+    {
+        return powInternal(t);
+    }
 
-    FloatType& pow(float f); 
-    FloatType& pow(const FloatType&);
-    FloatType& pow(const DoubleType&);
-    FloatType& pow(const IntType&);
+    Numeric& apply(std::function<Numeric&(Type&)> func)
+    {
+        if (func) 
+        {    
+            return func(*value);
+        }
+    
+        return *this;
+    }
 
-    FloatType& apply(std::function<FloatType&(float&)>);
-    FloatType& apply(void(*)(float&));
+    Numeric& apply(void(*ptrFunc)(Type&))
+    {
+        if(ptrFunc)
+        {
+            ptrFunc(*value);
+        }
+    
+        return *this;
+    }
 
 private:
-    float* value = nullptr;
-    FloatType& powInternal(const float f);
+    std::unique_ptr<Type> value;
+    Numeric& powInternal(const Type& t)
+    {
+        *value = static_cast<Type>(std::pow( *value, t ));
+        return *this;
+    }
 };
-
-struct DoubleType
+//----------------------------------------------
+template<>
+struct Numeric<double>
 {
-    explicit DoubleType(double val) : value(new double(val)) {}
-    ~DoubleType()
+    using Type = double;
+    explicit Numeric(Type val) : value(std::make_unique<Type>(val)) {}
+    ~Numeric()
     {
-        delete value;
+        value.reset(nullptr);
```

Comment:
This destructor is unnecessary.  
`unique_ptr` automatically destroys the owned object when it goes out of scope, which is what happens when the `Numeric<>` class is destroyed.
- [ ] revise